File: 62-unique-paths/62-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        @cache
        def helper(i: int, j: int) -> int:
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            
            if i == m - 1 and j == n - 1:
                return 1
            
            return helper(i + 1, j) + helper(i, j + 1)
        return helper(0, 0)

File: 62-unique-paths/62-unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        return this.helper(m, n, 0, 0, new int[m][n]);
    }
    
    private int helper(int m, int n, int row, int col, int[][] dp){
        if(row >= m || col >= n){
            return 0;
        }
        
        if(row == m - 1 && col == n - 1){
            return 1;
        }
        
        if(dp[row][col] == 0){
            dp[row][col] = helper(m, n, row + 1, col, dp) + helper(m, n, row, col + 1, dp);
        }
        
        return dp[row][col];
    }
}